## 5.  Improving model predictions through experimentation (hyperparameter tuning)

The first predictions you make with a model are generally referred to as baseline predictions. The same goes with the first evaluation metrics you get. These are generally referred to as baseline metrics.

Your next goal is to improve upon these baseline metrics.

Two of the main methods to improve baseline metrics are from a data perspective and a model perspective.

From a data perspective asks:
* Could we collect more data? In machine learning, more data is generally better, as it gives a model more opportunities to learn patterns.
* Could we improve our data? This could mean filling in missIng values or finding a better encoding (turning things into numbers) strategy.

From a model perspective asks:
* Is there a better model we could use? If you've started out with a simple model, could you use a more complex one? (we saw an example of this when looking at the [Scikit-Learn machine learning map](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html), ensemble methods are generally considered more complex models)
* Could we improve the current model? If the model you're using performs well straight out of the box, can the **hyperparameters** be tuned to make it even better?

**Note:** Patterns in data are also often referred to as data parameters. The difference between parameters and hyperparameters is a machine learning model seeks to find parameters in data on its own, where as, hyperparameters are settings on a model which a user (you) can adjust.

Since we have two existing datasets, we'll come at exploration from a model perspective.

More specifically, we'll look at how we could improve our `RandomForestClassifier` and `RandomForestRegressor` models through hyperparameter tuning.

What even are hyperparameters?

Good question, let's check it out. First, we'll instantiate a `RandomForestClassifier`.

Now let's starts with importing the data from CSV format to data frame.

In [1]:
# import numpy and pandas to import data from csv format
import numpy as np, pandas as pd
heart_df = pd.read_csv('../data/heart-disease.csv')
# test if the import is working
heart_df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [2]:
# check if there are NaN in the data frame
heart_df.isnull().sum().sum()

0

In [3]:
# Okay let's take a closer look
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [4]:
# next we separate the target to features
heart_y = heart_df['target']
# meanwhile the X are all of the features
heart_X = heart_df.drop('target', axis=1)
# check it out
heart_y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [5]:
heart_X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [6]:
# next we separate the X train, X test, y train and y test
# to do this we need to import sklearn.model_selection.train_test_split
from sklearn.model_selection import train_test_split

# next before we use the library we need to set our random seed to have consistent result
np.random.seed(42)

# use the train_test_split to split the current model with 80:20 for the moment as for train : test portion.

heart_X_train, heart_X_test, heart_y_train, heart_y_test = train_test_split(heart_X, heart_y, test_size=.2)

In [7]:
# Check the data table shapes to ensure correct proportions
(heart_X_train.shape, heart_y_train.shape, heart_X_test.shape, heart_y_test.shape)

((242, 13), (242,), (61, 13), (61,))

As seen here the table shape is consistent as the train has 242 rows and the test has 61 rows

In [8]:
# Begin creating a model based on the sklearn.ensmeble.RandomForestCalssiier
from sklearn.ensemble import RandomForestClassifier
# put it into easier to write object
rfc = RandomForestClassifier()

rfc.fit(heart_X_train, heart_y_train)
rfc.score(heart_X_test, heart_y_test)

0.8524590163934426

Thus, the score parameter conclude the preparations. Next we will go to the Hyperparameters Tuning Section

### See available parameters
When setting the model algorithm using Random Forest Classifier above you are using the default hyperparameters. 

Now let's see all available parameters for the Random Forest Classifier: 

We call the model instance (`rfc` in this case) and use `get_params()`

In [9]:
# NOTE: get_params is a function thus to get the return not the object you need to add ():
rfc.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

You'll see things like `max_depth`, `min_samples_split`, `n_estimators`.

Each of these is a hyperparameter of the `RandomForestClassifier` you can adjsut.

You can think of hyperprameters as being similar to dials on a over. Ont he default setting your oven might do an okay job cooking your favorite meal. But with a little experimentation, you find it does better when you adjust the settings.

<img src="../images/sklearn-hyperparameter-tuning-oven.png" width=400/>

The same goes for improving a machine learning model by hyperparameter tuning. The default hyperparameters on a machine learning model may find patterns in data well. But, there's a chance by adjusting the hyperparameters may improve a model's performace.

Every machine learning model will have different hyperparameter you can tune.

Instead of memorizing all of the hyperparameters for every model, we'll see how it's done with one. And then knowing these principles, you can apply them to a different model if needed.

Reading the <a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html">Scikit-Learn documentation for the Random Forest</a>, you'll find they suggest trying to change `n_estimators` (the number of trees in the forest) and `min_samples_split` (the minimum number of samples required to split and internal node).

We'll try tuning these as well as:
* `max_features` (the number of features to consider when looking for the best split)
* `max_depth` (the maximum depth of the tree)
* `min_samples_leaf` (the minimum number of samples required to be at a lea node)

If this still sound like a lot, the good news is, the process we're talking with the Random Forest and tuning its hyperparameters, can be used for other machine learning models in Scikit-Learn. The only difference is, with a different model, the hyperparameters you tune will be different.

Adjusting hyperparameters is usually an experimental process to figure out which are best. As there's no real way of knowing which hyperparemters will be best when starting out.

To get familiar with hyperparameter tuning, we'll take our Random Forest Classifier and adjust its hyperparameters in 3 ways:
1. By hand
1. Randomly with <a href= "https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html">RandomizedSearchCV</a>
1. Exhaustively with <a href="https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html"> GridSearchCV </a>

### 5.1 Tuning hyperparameters by hand

So far we've worked with training and test datasets.

You train a model on a training set and evaluate it on a test dataset.

But hyperparameter tuning introduces a third set, a validation set.

Now the process becomes: 
- train a model on the training data, 
- (try to) improve its hyperparameters on the validation set 
- evaluate it on the test dataset.

If our starting dataset contained 100 different patient records labels indicating who had heart disease and who didn't and we wanted to build a machine learning model to predict who had heart disease and who didn't, it might look like this:

<img src="../images/sklearn-train-valid-test-annotated.png" width=500/>

Since we know we're using a `RandomForestClassifier` and we know the hyperparameters we want to adjust, let's see what it looks like.

First, let's remind ourselves of the base parameters.

In [10]:
rfc.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

And we're going to adjust:
* `max_depth`
* `max_features`
* `min_samples_leaf`
* `min_samples_split`
* `n_estimators`

We'll use the same code as before, except this time we'll create a training, validation and test split.

With the training set containing 70% of the data and the validation and test sets each containing 15%.

Let's get some baseline results, then we'll tune the model.

And since we're going to be evaluating a few models, let's make an evaluation function.

In [11]:
# I need to import some libs from the sklearn.metrics to compare the hyperparameter tunings
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def evaluate_hyper(y_true, y_preds):
    """  
    Performs evaluation comparison on y_true vs y_preds labels.
    """
    # set the metrics used to evaluate the model performance after tuning
    accuracy = accuracy_score(y_true, y_preds)
    precision = precision_score(y_true, y_preds)
    recall = recall_score(y_true, y_preds)
    f1 = f1_score(y_true, y_preds)

    # put all of them into dictionary which make it easier to call and evaluate later on
    metric_dict = {"accuracy" : round(accuracy, 2), 
                    "precision" : round(precision, 2),
                    "recall" : round(recall, 2),
                    "f1" : round(f1, 2)}
    
    # print the result (for testing meaybe):
    print(f"Acc: {accuracy * 100:.2f}%")
    print(f"Precision: {precision:.2f}")
    print(f"Recall: {recall:.2f}")
    print(f"F1 score: {f1:.2f}")

    return metric_dict
    

In [13]:
# Now let's run the scoring to provide comparison parameters
# we already import accuracy_score, precision_score, recall_score, f1_score
# rfc = RandomForestClassifiers
# re-seed the random generator
np.random.seed(42)

# shuffle the data
heart_df = heart_df.sample(frac=1)
# more about pandas.dataframe sample Methods: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html#pandas.DataFrame.sample

# Basically we are picking random sample from the heart_df but with frac = 1 we take all data as sample but at random
# this is the same as randomizing the original heart_df data frame.

# Now we re-doing the same steps on modelling data:
# Split into X and y:
heart_X = heart_df.drop("target", axis=1)
heart_y = heart_df["target"]

# Now we need to separate them into : train, validation, and test data
# we use the set train = 70% of data, validation = 15% of data and test = 15% of data
train_split = round(0.7 * len(heart_df)) # 70 % of the whole data: NOTE: we only take the length NOT the data
validation_split = round(train_split + 0.15 * len(heart_df)) # 15% of the whole length of heart_df used for validation

# use those split numbers to pick the sampele from X and y (data from index 0 to train_split)
heart_X_train, heart_y_train = heart_X[:train_split], heart_y[:train_split]
# for the validation data remember the index starts from train_split to validation split
heart_X_validation, heart_y_validation = heart_X[train_split : validation_split], heart_y[train_split : validation_split]
# thus the rest (index validation_split to end) is the test part:
heart_X_test, heart_y_test = heart_X[validation_split:], heart_y[validation_split:]

# now we will use the validation to test the model developed from train portion of the data
rfc.fit(heart_X_train, heart_y_train)
# now we make prediction with the fitted rfc:
heart_y_pred = rfc.predict(heart_X_validation)

# Now we have the heart_y_pred which resulted from model RFC so we can score it against the heart_y_validation
# since we use heart_X_validation portion to make these predictions:
# Oh to make all process faster we already make function evaluate_hyper to print out the result
heart_baseline_metrics = evaluate_hyper(heart_y_validation, heart_y_pred)
heart_baseline_metrics

Acc: 84.44%
Precision: 0.87
Recall: 0.90
F1 score: 0.89


{'accuracy': 0.84, 'precision': 0.87, 'recall': 0.9, 'f1': 0.89}

Now we have the **Baseline metrics** as parameter that can be compared when we make changes into hyperparameter.

Let's start with tuning the most common hyperparamter in Random Forest Classifier which is number estimator (tree in the forest) (n_estimator), find out more on [sklearn.ensamble.RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)


In [14]:
# we already import sklear.ensemble.RandomforestClassiier 
# we make the second instance model of this and name it rfc_2 to differentiate prior rfc

rfc_2 = RandomForestClassifier(n_estimators=100)
# we use n_estimator as 100

# using the train and validation data we crate the model 
rfc_2.fit(heart_X_train, heart_y_train)
# using this model to make prediction using heart_X_validation data:
heart_y_pred_2 = rfc_2.predict(heart_X_validation)

# evaluate using the 2nd RFC model:
rfc_2_metrics = evaluate_hyper(heart_y_validation, heart_y_pred_2)
rfc_2_metrics

Acc: 80.00%
Precision: 0.86
Recall: 0.83
F1 score: 0.85


{'accuracy': 0.8, 'precision': 0.86, 'recall': 0.83, 'f1': 0.85}

The result is slightly bad, but the journey is still long, there are still many value for n_estimator.

Also how about other hyperparamters ready to be tuned?

Also we haven't use the test portion of the data....

However, one thing is clear... using this kind of manual by hand workflow is not practical. We need better method to tune broad range of hyperparameters.

### 5.2 Hyperparameter tunign with [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)

Scikit-Learn's [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html) allows us to randomly search accross different hyperparameters to see which work best. It also stores details about the ones which work best!

Let's see it in action

**PAY ATTENTION EACH COMMENT WHICH HAS STEP BY STEP PROCEDURES**

In [15]:
# TODO: continue to section 5.2